File: artifacts/api-server/src/services/polymarket.ts
import { logger } from "../lib/logger";

const GAMMA_API = "https://gamma-api.polymarket.com";

export interface GammaMarket {
  id: string;
  question: string;
  conditionId: string;
  slug: string;
  startDate: string;
  endDate: string;
  description: string;
  outcomes: string;
  outcomePrices: string;
  volume: string;
  active: boolean;
  closed: boolean;
  archived: boolean;
  volume24hr: number;
  liquidity: number;
  clobTokenIds?: string;
  tags?: Array<{ id: string; label: string; slug: string }>;
  resolvedBy?: string;
  resolutionSource?: string;
  image?: string;
}

export interface NormalizedMarket {
  id: string;
  question: string;
  category: string;
  status: "active" | "resolved" | "closed";
  yesPrice: number;
  noPrice: number;
  volume: number;
  volume24h: number;
  liquidity: number;
  endDate: string;
  resolvedOutcome: string | null;
  description: string;
  conditionId: string;
  tokenId: string;
}

function parseOutcomePrices(raw: string): [number, number] {
  try {
    const arr = JSON.parse(raw);
    return [parseFloat(arr[0]) || 0.5, parseFloat(arr[1]) || 0.5];
  } catch {
    return [0.5, 0.5];
  }
}

function parseTokenId(raw?: string): string {
  if (!raw) return "";
  try {
    const arr = JSON.parse(raw);
    return Array.isArray(arr) && arr[0] ? String(arr[0]) : "";
  } catch {
    return "";
  }
}

function parseCategory(tags?: GammaMarket["tags"]): string {
  if (!tags || tags.length === 0) return "General";
  const label = tags[0].label;
  return label.charAt(0).toUpperCase() + label.slice(1);
}

export function normalizeMarket(m: GammaMarket): NormalizedMarket {
  const [yesPrice, noPrice] = parseOutcomePrices(m.outcomePrices);
  const status: NormalizedMarket["status"] = m.closed || m.archived
    ? "resolved"
    : m.active
    ? "active"
    : "closed";

  return {
    id: m.id,
    question: m.question,
    category: parseCategory(m.tags),
    status,
    yesPrice,
    noPrice,
    volume: parseFloat(m.volume) || 0,
    volume24h: m.volume24hr || 0,
    liquidity: m.liquidity || 0,
    endDate: m.endDate,
    resolvedOutcome: null,
    description: m.description || "",
    conditionId: m.conditionId || "",
    tokenId: parseTokenId(m.clobTokenIds),
  };
}

async function fetchWithTimeout(url: string, timeoutMs = 10000): Promise<Response> {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, { signal: controller.signal });
    return res;
  } finally {
    clearTimeout(id);
  }
}

async function fetchWithRetry(url: string, retries = 3, delayMs = 600): Promise<Response> {
  let lastError: unknown;
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      const res = await fetchWithTimeout(url);
      if (res.ok) return res;
      if (res.status >= 400 && res.status < 500) {
        throw new Error(`Gamma API client error: ${res.status} ${res.statusText}`);
      }
      lastError = new Error(`Gamma API server error: ${res.status}`);
      logger.warn({ url, attempt, status: res.status }, "Gamma API non-ok, retrying");
    } catch (e) {
      lastError = e;
      logger.warn({ url, attempt, err: e }, "Gamma API fetch failed, retrying");
    }
    if (attempt < retries) {
      await new Promise((r) => setTimeout(r, delayMs * attempt));
    }
  }
  throw lastError;
}

export async function fetchMarkets(params?: {
  active?: boolean;
  limit?: number;
  offset?: number;
}): Promise<GammaMarket[]> {
  const qs = new URLSearchParams({
    limit: String(params?.limit ?? 100),
    offset: String(params?.offset ?? 0),
    ...(params?.active !== undefined ? { active: String(params.active) } : {}),
  });

  const url = `${GAMMA_API}/markets?${qs}`;
  logger.info({ url }, "Fetching markets from Gamma API");

  const res = await fetchWithRetry(url);
  const data = await res.json() as GammaMarket[];
  return Array.isArray(data) ? data : [];
}

export async function fetchAllActiveMarkets(maxPages = 5): Promise<GammaMarket[]> {
  const pageSize = 200;
  const all: GammaMarket[] = [];

  for (let page = 0; page < maxPages; page++) {
    const offset = page * pageSize;
    logger.info({ page, offset }, "Fetching page from Gamma API");
    try {
      const batch = await fetchMarkets({ active: true, limit: pageSize, offset });
      if (batch.length === 0) break;
      all.push(...batch);
      if (batch.length < pageSize) break;
    } catch (e) {
      logger.warn({ page, err: e }, "Gamma API page fetch failed, stopping pagination");
      break;
    }
  }

  return all;
}

export async function fetchMarketById(id: string): Promise<GammaMarket | null> {
  const url = `${GAMMA_API}/markets/${id}`;
  try {
    const res = await fetchWithRetry(url, 2);
    if (!res.ok) return null;
    return await res.json() as GammaMarket;
  } catch {
    return null;
  }
}

let _cache: { markets: NormalizedMarket[]; ts: number } | null = null;
const CACHE_TTL = 5 * 60 * 1000;

export async function getCachedMarkets(): Promise<NormalizedMarket[]> {
  if (_cache && Date.now() - _cache.ts < CACHE_TTL) {
    return _cache.markets;
  }
  const raw = await fetchAllActiveMarkets(5);
  const markets = raw
    .filter((m) => !m.archived)
    .map(normalizeMarket);
  _cache = { markets, ts: Date.now() };
  logger.info({ count: markets.length }, "Markets cache refreshed");
  return markets;
}

export function invalidateCache() {
  _cache = null;
}

File: server/internal/services/polymarket.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"polymarket-trader/internal/models"
)

const gammaAPI = "https://gamma-api.polymarket.com"
const cacheTTL = 5 * time.Minute
const significantChangePct = 2.0

type marketCache struct {
	mu         sync.RWMutex
	markets    []models.NormalizedMarket
	ts         time.Time
	prevPrices map[string]float64
}

var mktCache = &marketCache{prevPrices: make(map[string]float64)}

type gammaMarket struct {
	ID            string      `json:"id"`
	Question      string      `json:"question"`
	ConditionID   string      `json:"conditionId"`
	EndDate       string      `json:"endDate"`
	Description   string      `json:"description"`
	OutcomePrices string      `json:"outcomePrices"`
	Volume        string      `json:"volume"`
	Active        bool        `json:"active"`
	Closed        bool        `json:"closed"`
	Archived      bool        `json:"archived"`
	Volume24hr    interface{} `json:"volume24hr"`
	Liquidity     interface{} `json:"liquidity"`
	ClobTokenIds  *string     `json:"clobTokenIds"`
	Tags          []struct {
		Label string `json:"label"`
	} `json:"tags"`
}

func parseFloat(v interface{}) float64 {
	switch val := v.(type) {
	case float64:
		return val
	case string:
		f, _ := strconv.ParseFloat(val, 64)
		return f
	}
	return 0
}

func parseOutcomePrices(raw string) (float64, float64) {
	var arr []string
	if err := json.Unmarshal([]byte(raw), &arr); err != nil || len(arr) < 2 {
		return 0.5, 0.5
	}
	yes, _ := strconv.ParseFloat(arr[0], 64)
	no, _ := strconv.ParseFloat(arr[1], 64)
	if yes == 0 {
		yes = 0.5
	}
	if no == 0 {
		no = 0.5
	}
	return yes, no
}

func parseTokenID(raw *string) string {
	if raw == nil {
		return ""
	}
	var arr []interface{}
	if err := json.Unmarshal([]byte(*raw), &arr); err != nil || len(arr) == 0 {
		return ""
	}
	if s, ok := arr[0].(string); ok {
		return s
	}
	return ""
}

func parseCategory(tags []struct{ Label string `json:"label"` }) string {
	if len(tags) == 0 {
		return "General"
	}
	l := tags[0].Label
	if len(l) == 0 {
		return "General"
	}
	return strings.ToUpper(l[:1]) + l[1:]
}

func parseDateSafe(raw string) string {
	if raw == "" {
		return time.Unix(0, 0).UTC().Format(time.RFC3339)
	}
	t, err := time.Parse(time.RFC3339, raw)
	if err != nil {
		t, err = time.Parse("2006-01-02T15:04:05Z", raw)
		if err != nil {
			return time.Unix(0, 0).UTC().Format(time.RFC3339)
		}
	}
	return t.UTC().Format(time.RFC3339)
}

func NormalizeMarket(m gammaMarket) models.NormalizedMarket {
	yes, no := parseOutcomePrices(m.OutcomePrices)
	status := "closed"
	if m.Closed || m.Archived {
		status = "resolved"
	} else if m.Active {
		status = "active"
	}

	vol, _ := strconv.ParseFloat(m.Volume, 64)
	return models.NormalizedMarket{
		ID:              m.ID,
		Question:        m.Question,
		Category:        parseCategory(m.Tags),
		Status:          status,
		YesPrice:        yes,
		NoPrice:         no,
		Volume:          vol,
		Volume24h:       parseFloat(m.Volume24hr),
		Liquidity:       parseFloat(m.Liquidity),
		EndDate:         parseDateSafe(m.EndDate),
		ResolvedOutcome: nil,
		Description:     m.Description,
		ConditionID:     m.ConditionID,
		TokenID:         parseTokenID(m.ClobTokenIds),
	}
}

func fetchWithTimeout(rawURL string) ([]byte, int, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	req, _ := http.NewRequestWithContext(ctx, "GET", rawURL, nil)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, resp.StatusCode, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var buf []byte
	tmp := make([]byte, 32*1024)
	for {
		n, err := resp.Body.Read(tmp)
		buf = append(buf, tmp[:n]...)
		if err != nil {
			break
		}
	}
	return buf, resp.StatusCode, nil
}

func fetchMarkets(active bool, limit, offset int) ([]gammaMarket, error) {
	params := url.Values{}
	params.Set("limit", strconv.Itoa(limit))
	params.Set("offset", strconv.Itoa(offset))
	if active {
		params.Set("active", "true")
	}
	u := fmt.Sprintf("%s/markets?%s", gammaAPI, params.Encode())

	var lastErr error
	for attempt := 1; attempt <= 3; attempt++ {
		body, code, err := fetchWithTimeout(u)
		if err == nil && code < 400 {
			var markets []gammaMarket
			if err2 := json.Unmarshal(body, &markets); err2 != nil {
				return nil, err2
			}
			return markets, nil
		}
		lastErr = err
		if code >= 400 && code < 500 {
			break
		}
		if attempt < 3 {
			time.Sleep(time.Duration(attempt*600) * time.Millisecond)
		}
	}
	return nil, lastErr
}

func fetchAllActiveMarkets() ([]gammaMarket, error) {
	pageSize := 200
	var all []gammaMarket
	for page := 0; page < 5; page++ {
		batch, err := fetchMarkets(true, pageSize, page*pageSize)
		if err != nil {
			log.Printf("Gamma API page %d failed: %v", page, err)
			break
		}
		if len(batch) == 0 {
			break
		}
		all = append(all, batch...)
		if len(batch) < pageSize {
			break
		}
	}
	return all, nil
}

func FetchMarketByID(id string) (*models.NormalizedMarket, error) {
	u := fmt.Sprintf("%s/markets/%s", gammaAPI, id)
	for attempt := 1; attempt <= 2; attempt++ {
		body, _, err := fetchWithTimeout(u)
		if err != nil {
			continue
		}
		var gm gammaMarket
		if err2 := json.Unmarshal(body, &gm); err2 != nil {
			return nil, err2
		}
		nm := NormalizeMarket(gm)
		return &nm, nil
	}
	return nil, fmt.Errorf("market not found")
}

func GetCachedMarkets() ([]models.NormalizedMarket, error) {
	mktCache.mu.RLock()
	if !mktCache.ts.IsZero() && time.Since(mktCache.ts) < cacheTTL {
		markets := mktCache.markets
		mktCache.mu.RUnlock()
		return markets, nil
	}
	mktCache.mu.RUnlock()

	mktCache.mu.Lock()
	defer mktCache.mu.Unlock()

	if !mktCache.ts.IsZero() && time.Since(mktCache.ts) < cacheTTL {
		return mktCache.markets, nil
	}

	raw, err := fetchAllActiveMarkets()
	if err != nil {
		return nil, err
	}

	var markets []models.NormalizedMarket
	for _, m := range raw {
		if !m.Archived {
			markets = append(markets, NormalizeMarket(m))
		}
	}

	newPrices := make(map[string]float64)
	var movers int
	for _, m := range markets {
		newPrices[m.ID] = m.YesPrice
		if prev, ok := mktCache.prevPrices[m.ID]; ok && prev > 0 {
			chg := math.Abs((m.YesPrice-prev)/prev) * 100
			if chg >= significantChangePct {
				movers++
			}
		}
	}

	log.Printf("Markets cache refreshed: total=%d significant_movers=%d", len(markets), movers)
	mktCache.markets = markets
	mktCache.ts = time.Now()
	mktCache.prevPrices = newPrices
	return markets, nil
}

func InvalidateCache() {
	mktCache.mu.Lock()
	defer mktCache.mu.Unlock()
	mktCache.ts = time.Time{}
}

func GetCachedPrevPrices() map[string]float64 {
	mktCache.mu.RLock()
	defer mktCache.mu.RUnlock()
	cp := make(map[string]float64, len(mktCache.prevPrices))
	for k, v := range mktCache.prevPrices {
		cp[k] = v
	}
	return cp
}
